File: tools/split_functions.py
import os
import re

BASE = r"c:\xampp\htdocs\coinrex\includes"
SRC_PATH = os.path.join(BASE, "functions.php")


def extract_functions(lines):
    funcs = []
    outside = []
    i = 0
    n = len(lines)

    while i < n:
        line = lines[i]
        if re.match(r"^\s*function\s+[A-Za-z_][A-Za-z0-9_]*\s*\(", line):
            start = i
            name = re.findall(r"function\s+([A-Za-z_][A-Za-z0-9_]*)\s*\(", line)[0]
            brace = 0
            started = False
            while i < n:
                l = lines[i]
                brace += l.count("{")
                if "{" in l:
                    started = True
                brace -= l.count("}")
                if started and brace == 0:
                    end = i
                    i += 1
                    break
                i += 1
            block = "\n".join(lines[start : end + 1])
            funcs.append((name, block))
        else:
            outside.append(line)
            i += 1
    return funcs, outside


def mod_for(name):
    lname = name.lower()
    if lname.startswith("taskhub") or "taskhub" in lname:
        return "taskhub"
    if "boosthub" in lname:
        return "boosthub"
    if lname.startswith("send") or "otp" in lname or "mail" in lname:
        return "email"
    if "remember" in lname or lname in [
        "loginuser",
        "logoutuser",
        "isloggedin",
        "establishauthenticatedsession",
        "restorerememberedsession",
        "touchauthenticateduseractivity",
    ]:
        return "auth"
    if lname.startswith("getuser") or lname.startswith("updateuser") or lname in [
        "registeruser",
        "uploadprofileavatar",
        "isuserprofilecomplete",
        "getcurrentuser",
        "getcurrentuserid",
        "generateusername",
    ]:
        return "user"
    if "ledger" in lname or "claim" in lname or "reward" in lname:
        return "reward_ledger"
    if "referral" in lname:
        return "referrals"
    if "level" in lname or "review" in lname or "project" in lname or "expert" in lname:
        return "levels"
    if (
        "csrf" in lname
        or "sanitize" in lname
        or "security" in lname
        or "disposable" in lname
        or "passwordpolicy" in lname
        or "clientip" in lname
    ):
        return "security"
    if lname in ["redirect", "normalizeemail", "normalizereferralcode", "getemaildomain"]:
        return "helpers"
    return "core"


def main():
    with open(SRC_PATH, "r", encoding="utf-8") as f:
        content = f.read()

    raw = content.replace("<?php", "", 1)
    raw = raw.rsplit("?>", 1)[0]
    lines = raw.splitlines()
    funcs, outside = extract_functions(lines)

    modules = {
        "core": [],
        "auth": [],
        "user": [],
        "email": [],
        "reward_ledger": [],
        "taskhub": [],
        "boosthub": [],
        "levels": [],
        "referrals": [],
        "security": [],
        "helpers": [],
    }

    for name, block in funcs:
        modules[mod_for(name)].append(block)

    func_dir = os.path.join(BASE, "functions")
    os.makedirs(func_dir, exist_ok=True)
    outside_txt = "\n".join(outside).strip("\n")

    for mod, blocks in modules.items():
        path = os.path.join(func_dir, f"{mod}.php")
        with open(path, "w", encoding="utf-8") as f:
            f.write("<?php\n")
            f.write("/** Auto-split from legacy functions.php */\n\n")
            if mod == "core" and outside_txt.strip():
                f.write(outside_txt + "\n\n")
            f.write("\n\n".join(blocks).strip() + "\n")

    legacy = os.path.join(BASE, "functions_legacy_backup.php")
    if not os.path.exists(legacy):
        with open(legacy, "w", encoding="utf-8") as f:
            f.write("<?php\n" + raw + "\n?>")

    loader = """<?php
/**
 * CoinRex Helper Functions Loader
 * Split into modular files for maintainability.
 */

require_once __DIR__ . '/functions/core.php';
require_once __DIR__ . '/functions/helpers.php';
require_once __DIR__ . '/functions/security.php';
require_once __DIR__ . '/functions/user.php';
require_once __DIR__ . '/functions/email.php';
require_once __DIR__ . '/functions/reward_ledger.php';
require_once __DIR__ . '/functions/boosthub.php';
require_once __DIR__ . '/functions/taskhub.php';
require_once __DIR__ . '/functions/referrals.php';
require_once __DIR__ . '/functions/levels.php';
require_once __DIR__ . '/functions/auth.php';
?>
"""
    with open(SRC_PATH, "w", encoding="utf-8") as f:
        f.write(loader)

    print("Split complete. Functions:", len(funcs))
    for k, v in modules.items():
        print(k, len(v))


if __name__ == "__main__":
    main()

File: tools/split_functions.php
<?php

$base = 'c:/xampp/htdocs/coinrex/includes';
$srcPath = $base . '/functions.php';

if (!file_exists($srcPath)) {
    fwrite(STDERR, "functions.php not found\n");
    exit(1);
}

$content = file_get_contents($srcPath);
$content = preg_replace('/^<\?php\s*/', '', $content, 1);
$content = preg_replace('/\?>\s*$/', '', $content, 1);
$lines = preg_split('/\R/', $content);

$functions = [];
$outside = [];

$i = 0;
$n = count($lines);
while ($i < $n) {
    $line = $lines[$i];
    if (preg_match('/^\s*function\s+([A-Za-z_][A-Za-z0-9_]*)\s*\(/', $line, $m)) {
        $name = $m[1];
        $start = $i;
        $brace = 0;
        $started = false;
        while ($i < $n) {
            $l = $lines[$i];
            $brace += substr_count($l, '{');
            if (strpos($l, '{') !== false) {
                $started = true;
            }
            $brace -= substr_count($l, '}');
            if ($started && $brace === 0) {
                $end = $i;
                $i++;
                break;
            }
            $i++;
        }
        $block = implode("\n", array_slice($lines, $start, $end - $start + 1));
        $functions[] = [$name, $block];
    } else {
        $outside[] = $line;
        $i++;
    }
}

$modules = [
    'core' => [],
    'auth' => [],
    'user' => [],
    'email' => [],
    'reward_ledger' => [],
    'taskhub' => [],
    'boosthub' => [],
    'levels' => [],
    'referrals' => [],
    'security' => [],
    'helpers' => [],
];

$modFor = function ($name) {
    $lname = strtolower($name);
    if (strpos($lname, 'taskhub') !== false) return 'taskhub';
    if (strpos($lname, 'boosthub') !== false) return 'boosthub';
    if (strpos($lname, 'otp') !== false || strpos($lname, 'mail') !== false || strpos($lname, 'smtp') !== false) return 'email';
    if (strpos($lname, 'remember') !== false || in_array($lname, ['loginuser','logoutuser','isloggedin','establishauthenticatedsession','restorerememberedsession','touchauthenticateduseractivity'], true)) return 'auth';
    if (strpos($lname, 'getuser') === 0 || strpos($lname, 'updateuser') === 0 || in_array($lname, ['registeruser','uploadprofileavatar','isuserprofilecomplete','getcurrentuser','getcurrentuserid','generateusername'], true)) return 'user';
    if (strpos($lname, 'ledger') !== false || strpos($lname, 'claim') !== false || strpos($lname, 'reward') !== false) return 'reward_ledger';
    if (strpos($lname, 'referral') !== false) return 'referrals';
    if (strpos($lname, 'level') !== false || strpos($lname, 'review') !== false || strpos($lname, 'project') !== false || strpos($lname, 'expert') !== false) return 'levels';
    if (strpos($lname, 'csrf') !== false || strpos($lname, 'sanitize') !== false || strpos($lname, 'security') !== false || strpos($lname, 'disposable') !== false || strpos($lname, 'passwordpolicy') !== false || strpos($lname, 'clientip') !== false) return 'security';
    if (in_array($lname, ['redirect','normalizeemail','normalizereferralcode','getemaildomain'], true)) return 'helpers';
    return 'core';
};

foreach ($functions as [$name, $block]) {
    $modules[$modFor($name)][] = $block;
}

$funcDir = $base . '/functions';
if (!is_dir($funcDir)) {
    mkdir($funcDir, 0755, true);
}

$outsideTxt = trim(implode("\n", $outside));
foreach ($modules as $module => $blocks) {
    $path = $funcDir . '/' . $module . '.php';
    $body = "<?php\n/** Auto-split from legacy functions.php */\n\n";
    if ($module === 'core' && $outsideTxt !== '') {
        $body .= $outsideTxt . "\n\n";
    }
    $body .= trim(implode("\n\n", $blocks)) . "\n";
    file_put_contents($path, $body);
}

$legacy = $base . '/functions_legacy_backup.php';
if (!file_exists($legacy)) {
    file_put_contents($legacy, "<?php\n" . $content . "\n?>\n");
}

$loader = <<<'PHP'
<?php
/**
 * CoinRex Helper Functions Loader
 * Split into modular files for maintainability.
 */

require_once __DIR__ . '/functions/core.php';
require_once __DIR__ . '/functions/helpers.php';
require_once __DIR__ . '/functions/security.php';
require_once __DIR__ . '/functions/user.php';
require_once __DIR__ . '/functions/email.php';
require_once __DIR__ . '/functions/reward_ledger.php';
require_once __DIR__ . '/functions/boosthub.php';
require_once __DIR__ . '/functions/taskhub.php';
require_once __DIR__ . '/functions/referrals.php';
require_once __DIR__ . '/functions/levels.php';
require_once __DIR__ . '/functions/auth.php';
?>
PHP;

file_put_contents($srcPath, $loader . "\n");

echo "Split complete. Functions: " . count($functions) . PHP_EOL;
foreach ($modules as $k => $v) {
    echo $k . ' ' . count($v) . PHP_EOL;
}
